### PR TITLE
feat: allow round trip middleware per operation

### PR DESCRIPTION
**Pull request by pseudo-su**
Reference implementation to cater to the use-cases I described in #150

## Notes

Fixes https://github.com/deepmap/oapi-codegen/issues/150

### CODE REVIEW

**Comment by StevenACoffman** [comment]
This is exactly the use case I have, so I would love to see a complete example in your pseudo-su /golang-service-template repo

**Comment by pseudo-su** [comment]
@StevenACoffman I haven't forgotten about this, you make a good point about examples... I might actually add some more tests cases to this PR that would also help somewhat in capturing examples.

I have a few open PRs for `oapi-codegen` that may or may not get merged. I was trying to avoid it because I'm lazy but I might merge them all into my forked repo and add a branch to `pseudo-su /golang-service-template` with examples.

My preference was to get things merged (where possible) before spending too much time messing about with a temporary fork.

File: pkg/codegen/templates/client.tmpl [hunk 1]
```diff
@@ -42,9 +96,71 @@ func NewClient(server string, opts ...ClientOption) (*Client, error) {
     if client.Client == nil {
         client.Client = http.DefaultClient
     }
+
+    client.operationDoers = setupoperationDoers(&client, client.RoundTripMiddlewares)
+
     return &client, nil
 }
 
+func setupoperationDoers(c *Client, rtMiddlewares RoundTripMiddlewares) *operationDoFunctions {
```
Comment (StevenACoffman) [inline pkg/codegen/templates/client.tmpl:105]: The case on this is weird. 
```suggestion
func setupOperationDoers(c *Client, rtMiddlewares RoundTripMiddlewares) *operationDoFunctions {
```

File: pkg/codegen/templates/client.tmpl [hunk 2]
```diff
@@ -42,9 +96,71 @@ func NewClient(server string, opts ...ClientOption) (*Client, error) {
     if client.Client == nil {
         client.Client = http.DefaultClient
     }
+
+    client.operationDoers = setupoperationDoers(&client, client.RoundTripMiddlewares)
```
Comment (StevenACoffman) [inline pkg/codegen/templates/client.tmpl:100]:
```suggestion
    client.operationDoers = setupOperationDoers(&client, client.RoundTripMiddlewares)
```

**Comment by StevenACoffman** [comment]
@deepmap-marcinr This one is ready too!

**Comment by StevenACoffman** [comment]
@pseudo-su Hmmm, I think your rebase you need to regenerate because now there's some test failures.

**Comment by pseudo-su** [comment]
I need to fix some things and add those example/tests I mentioned

**Comment by pseudo-su** [comment]
Ok @StevenACoffman three main things fixed and this should be good now.

* joinMiddleware had an issue when passing an empty list of middleware to join.
* I've added tests in `internal/test/client/client_test.go` that test an "interceptor" using `WithSharedRoundTripMiddleware` and `WithRoundTripMiddleware`
* I fixed the spec in `internal/test/client/client.yaml` because the responses were missing the `content` field inbetween the status code and the content-type

EG this
```yaml
responses:
  200:
    application/json:
```
instead of 
```yaml
responses:
  200:
    content:
      application/json:
```

**Comment by StevenACoffman** [comment]
Thanks, yes, this works for us in our client. Nice improvement! @deepmap-marcinr Ready for you!

**Comment by batazor** [comment]
What is the status of this PR? Is there any other way to set custom behavior?

@StevenACoffman @pseudo-su

**Comment by DAcodedBEAT** [comment]
@oapi-codegen/oapi-codegen any reason why this wasn't approved/merged (aside from the current merge conflicts)?

**Comment by DAcodedBEAT** [comment]
@oapi-codegen any reason why this wasn't approved/merged (aside from the current merge conflicts)?

**Pull request closed without merging**